src/include/mpir_hw_topo.h

```diff
@@ -0,0 +1,232 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ */
+
+#ifndef MPIR_HW_TOPO_H_INCLUDED
+#define MPIR_HW_TOPO_H_INCLUDED
+
+/*
+ * Node hardware object types
+ */
+typedef enum {
+    MPIR_NODE_OBJ_TYPE__NONE = -1,
+    MPIR_NODE_OBJ_TYPE__MACHINE,
+    MPIR_NODE_OBJ_TYPE__PACKAGE,
+    MPIR_NODE_OBJ_TYPE__CORE,
+    MPIR_NODE_OBJ_TYPE__PU,
+    MPIR_NODE_OBJ_TYPE__L1CACHE,
+    MPIR_NODE_OBJ_TYPE__L2CACHE,
+    MPIR_NODE_OBJ_TYPE__L3CACHE,
+    MPIR_NODE_OBJ_TYPE__L4CACHE,
+    MPIR_NODE_OBJ_TYPE__L5CACHE,
+    MPIR_NODE_OBJ_TYPE__L1ICACHE,
+    MPIR_NODE_OBJ_TYPE__L2ICACHE,
+    MPIR_NODE_OBJ_TYPE__L3ICACHE,
+    MPIR_NODE_OBJ_TYPE__GROUP,
+    MPIR_NODE_OBJ_TYPE__NUMANODE,
+    MPIR_NODE_OBJ_TYPE__BRIDGE,
+    MPIR_NODE_OBJ_TYPE__PCI_DEVICE,
+    MPIR_NODE_OBJ_TYPE__OS_DEVICE,
+    MPIR_NODE_OBJ_TYPE__MISC,
+    MPIR_NODE_OBJ_TYPE__MAX
+} MPIR_Node_obj_type;
+
+/*
+ * Node hardware OS device object types
+ */
+typedef enum {
+    MPIR_NODE_OBJ_OSDEV_TYPE__NONE = -1,
+    MPIR_NODE_OBJ_OSDEV_TYPE__BLOCK,
+    MPIR_NODE_OBJ_OSDEV_TYPE__GPU,
+    MPIR_NODE_OBJ_OSDEV_TYPE__NETWORK,
+    MPIR_NODE_OBJ_OSDEV_TYPE__OPENFABRICS,
+    MPIR_NODE_OBJ_OSDEV_TYPE__DMA,
+    MPIR_NODE_OBJ_OSDEV_TYPE__COPROC
+} MPIR_Node_obj_osdev_type;
+
+/*
+ * Network topology types
+ */
+typedef enum {
+    MPIR_NETWORK_TOPOLOGY_TYPE__FAT_TREE,
+    MPIR_NETWORK_TOPOLOGY_TYPE__CLOS_NETWORK,
+    MPIR_NETWORK_TOPOLOGY_TYPE__TORUS,
+    MPIR_NETWORK_TOPOLOGY_TYPE__INVALID
+} MPIR_Network_topology_type;
+
+/*
+ * Network node types
+ */
+typedef enum {
+    MPIR_NETWORK_NODE_TYPE__HOST,
+    MPIR_NETWORK_NODE_TYPE__SWITCH,
+    MPIR_NETWORK_NODE_TYPE__INVALID
+} MPIR_Network_node_type;
+
+/*
+ * Definitions for node objects
+ */
+typedef void MPIR_Node_obj;
+
+/*
+ * Definitions for network objects
+ */
+typedef void MPIR_Network_node;
+typedef void MPIR_Network_edge;
+
```

Comment:
Just a question: are we constructing a graph here, that node is node and edge is edge, and `MPIR_Node_obj` are the attribute/value of the node?

Reply:
`MPIR_Node_obj` are node local objects that map to `hwloc`. `MPIR_Network_node/edge` are network objects that map to `netloc`. We construct network topology (graph) using `MPIR_Netloc_` functions in `MPIR_Hardware_topology_init`. When upper layers query the topology layer (either node or network) they get an opaque handle which they can use to get additional information from the topology layer. This way we don't need to expose any topology internal information to them. For example, you may want to get the id of the memory device covering the PU your process is running to bind memory to that device. In this case you query the topology layer which will return the opaque `MPIR_Node_obj` handle to the memory device; afterwards you can further query the the memory `id` by passing the such handle to another function.

Reply:
I see. Thanks.

Reply:
Why are these `void` instead of `void *`?  How can someone represent anything with a `void`?

Reply:
If you prefer I can define these as `void *` and change all the definitions from`MPIR_Node_obj *obj` to `MPIR_Node_obj obj`.

Reply:
I'm not sure that's a preference issue.  It's a correctness issue.  You can't represent anything with just `void`.  So you are working around the issue by using pointers in the calling functions.  For example, functions that need to return a node object, now have to return a pointer to a node object.

Reply:
I see, I will fix it

Reply:
Fixed

Reply:
If there is nothing else I missed I will rerun tests one more time